app/src/main/java/io/apicurio/registry/upgrade/DataUpgrader.java

```diff
@@ -0,0 +1,77 @@
+package io.apicurio.registry.upgrade;
+
+import io.apicurio.common.apps.config.Info;
+import io.apicurio.registry.storage.RegistryStorage;
+import io.apicurio.registry.storage.StorageEvent;
+import io.apicurio.registry.storage.StorageEventType;
+import io.apicurio.registry.storage.error.ReadOnlyStorageException;
+import io.apicurio.registry.storage.impexp.EntityInputStream;
+import io.apicurio.registry.types.Current;
+import io.apicurio.registry.utils.impexp.Entity;
+import io.apicurio.registry.utils.impexp.v2.EntityReader;
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.enterprise.event.ObservesAsync;
+import jakarta.inject.Inject;
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+import org.slf4j.Logger;
+
+import java.io.BufferedInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.Optional;
+import java.util.zip.ZipInputStream;
+
+@ApplicationScoped
+public class DataUpgrader {
+
+    @Inject
+    Logger log;
+
+    @Inject
+    @Current
+    RegistryStorage storage;
+
+    @ConfigProperty(name = "apicurio.upgrade.file.location")
+    @Info(category = "import", description = "The import URL", availableSince = "3.0.0")
+    Optional<URL> regisrtryV2ExportFile;
+
+    void onStorageReady(@ObservesAsync StorageEvent ev) {
+        if (StorageEventType.READY.equals(ev.getType()) && regisrtryV2ExportFile.isPresent()) {
+            log.info("Registry V2 export file exists.");
+            final URL registryV2ExportUrl = regisrtryV2ExportFile.get();
+            try (final InputStream registryV2ExportZip = new BufferedInputStream(
+                    registryV2ExportUrl.openStream())) {
+                log.info("Importing {} on startup.", registryV2ExportUrl);
```

Comment:
I can detect it here, but even if I don't it will fail down the path if it's not since all the entities will be parsed to v2 versions, so if it's not a V2 export zip it will fail to import the data. That said, it's probably better to fail fast here.

Reply:
Yeah that's what I was thinking:  fail fast.  Of course that assumes the manifest file is the first entry.  Which it is if the ZIP was created by Registry.